clvm-derive/src/from_clvm.rs

```diff
@@ -1,66 +1,309 @@
+            <#match_macro!( #( #field_types ),* )
+            as #crate_name::FromClvm<#node_name>>::from_clvm(decoder, node)?;
+        Ok(Self #initializer)
     };
 
-    let struct_name = &ast.ident;
-
-    // `match_macro` decodes a nested tuple containing each of the struct field types within.
-    // `destructure_macro` destructures the values into the field names, to be stored in the struct.
-    let (match_macro, destructure_macro) = match args.repr {
-        Repr::List => (
-            quote!( #crate_name::match_list ),
-            quote!( #crate_name::destructure_list ),
-        ),
-        Repr::Tuple => (
-            quote!( #crate_name::match_tuple ),
-            quote!( #crate_name::destructure_tuple ),
-        ),
-        Repr::Curry => (
-            quote!( #crate_name::match_curried_args ),
-            quote!( #crate_name::destructure_curried_args ),
-        ),
+    generate_from_clvm(crate_name, ast, &node_name, &body)
+}
+
+fn impl_for_enum(
+    crate_name: &TokenStream,
+    ast: &DeriveInput,
+    int_repr: &Ident,
+    variants: &[VariantInfo],
+) -> TokenStream {
+    let node_name = Ident::new("Node", Span::mixed_site());
+
+    let mut discriminant_definitions = Vec::new();
+    let mut has_initializers = false;
+
+    let variant_bodies = variants
+        .iter()
+        .enumerate()
+        .map(|(i, variant_info)| {
+            let VariantInfo {
+                name,
+                discriminant,
+                field_info,
+                macros,
+            } = variant_info;
+
+            let FieldInfo {
+                field_types,
+                field_names,
+                initializer,
+            } = field_info;
+
+            let Macros {
+                match_macro,
+                destructure_macro,
+                ..
+            } = macros;
+
+            let discriminant_ident = Ident::new(&format!("VALUE_{}", i), Span::mixed_site());
+            discriminant_definitions.push(quote! {
+                const #discriminant_ident: #int_repr = #discriminant;
+            });
+
+            if initializer.is_empty() {
+                quote! {
+                    #discriminant_ident => {
+                        Ok(Self::#name)
+                    }
+                }
+            } else {
+                has_initializers = true;
+                quote! {
+                    #discriminant_ident => {
+                        let #destructure_macro!( #( #field_names ),* ) =
+                            <#match_macro!( #( #field_types ),* )
+                            as #crate_name::FromClvm<#node_name>>::from_clvm(decoder, args.0)?;
+                        Ok(Self::#name #initializer)
+                    }
+                }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let parse_value = if has_initializers {
+        quote! {
+            let (value, args) = <(#int_repr, #crate_name::Raw<#node_name>)>::from_clvm(decoder, node)?;
```

Comment:
Both of these variables aren't used after the enum fields are brought into scope. I just tried it and it works fine currently, it just shadows the old variable and since the old one isn't used anywhere in that scope nothing breaks.